drivers/SmartThings/zigbee-smoke-detector/src/frient/init.lua

```diff
@@ -13,20 +13,254 @@
-  NAME = "Freint Smoke Detector",
+local function generate_event_from_zone_status(driver, device, zone_status, zigbee_message)
+
+   if zone_status:is_test_set() then
+      device:emit_event(smokeDetector.smoke.tested())
+   elseif zone_status:is_alarm1_set() then
+      device:emit_event(smokeDetector.smoke.detected())
+   else
+      device.thread:call_with_delay(6, function ()
+      device:emit_event(smokeDetector.smoke.clear())
+      end)
+   end
+end
+
+local function ias_zone_status_attr_handler(driver, device, zone_status, zb_rx)
+  generate_event_from_zone_status(driver, device, zone_status, zb_rx)
+end
+
+local function ias_zone_status_change_handler(driver, device, zb_rx)
+  local zone_status = zb_rx.body.zcl_body.zone_status
+  generate_event_from_zone_status(driver, device, zone_status, zb_rx)
+end
+
+local function send_siren_command(device)
+  -- Check firmware version first if not already known
+  local sw_version = device:get_field(PRIMARY_SW_VERSION)
+  if ((sw_version == nil) or (sw_version == "")) then
+    device:send(cluster_base.read_manufacturer_specific_attribute(device, Basic.ID, DEVELCO_BASIC_PRIMARY_SW_VERSION_ATTR, DEVELCO_MANUFACTURER_CODE))
+  end
+
+  local warning_duration = device:get_field(ALARM_LAST_DURATION) or DEFAULT_WARNING_DURATION
+  local sirenConfiguration
+  local warning_mode = 0x01  -- For siren on
+
+  if (device:get_field(SIREN_ENDIAN) == "reverse") then
+    -- Old frient firmware, the endian format is reversed
+    local siren_config_value = warning_mode
+    sirenConfiguration = IASWD.types.SirenConfiguration(siren_config_value)
+  else
+    sirenConfiguration = IASWD.types.SirenConfiguration(0x00)
+    sirenConfiguration:set_warning_mode(warning_mode)
+  end
+
+  device:send(
+    IASWD.server.commands.StartWarning(
+      device,
+      sirenConfiguration,
+      data_types.Uint16(warning_duration),
+      data_types.Uint8(00),
+      data_types.Enum8(00)
+    )
+  )
+end
+
+local emit_alarm_event = function(device, cmd)
+  if cmd == alarm_command.OFF then
+    device:emit_event(alarm.alarm.off())
+  else
+    if cmd == alarm_command.SIREN then
+      device:emit_event(alarm.alarm.siren())
+    end
+  end
+end
+
+local default_response_handler = function(driver, device, zigbee_message)
+  local command = zigbee_message.body.zcl_body.cmd.value
+  local alarm_ev = device:get_field(ALARM_COMMAND)
+  if command == IASWD.server.commands.StartWarning.ID then
+    if alarm_ev ~= alarm_command.OFF then
+      emit_alarm_event(device, alarm_ev)
+      local lastDuration = device:get_field(ALARM_LAST_DURATION) or ALARM_DEFAULT_MAX_DURATION
+      device.thread:call_with_delay(lastDuration, function(d)
+        device:emit_event(alarm.alarm.off())
+      end)
+    else
+      emit_alarm_event(device,alarm_command.OFF)
+    end
+  end
+end
+
+local siren_alarm_siren_handler = function(driver, device, command)
+  device:set_field(ALARM_COMMAND, alarm_command.SIREN, {persist = true})
```

Comment:
The `persist` here is probably unnecessary. Hub reboot will probably take longer than any alarm command and this info will be stale by the time the driver is running again